// office-addin-sso/lib/defaults.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.usageDataObject = exports.testManifestFilePath = exports.testFallbackAuthDialogFilePath = exports.testEnvDataFilePath = exports.getSecretCommandPath = exports.addSecretCommandPath = exports.manifestFilePath = exports.getInstalledAppsPath = exports.fallbackAuthDialogJavascriptFilePath = exports.fallbackAuthDialogTypescriptFilePath = exports.envDataFilePath = exports.azRestSetSigninAudienceCommandPath = exports.azRestSetIdentifierUriCommmandPath = exports.azRestGetTenantRolesPath = exports.azRestGetTenantAdminMembershipCommandPath = exports.azRestGetOrganizationDetailsCommandPath = exports.azRestAppCreateCommandPath = exports.azRestAddSecretCommandPath = exports.azRestAddTenantReplyUrlsCommandPath = exports.azRestAddTenantOutlookReplyUrlsCommandPath = exports.azCliInstallCommandPath = void 0;
const path_1 = __importDefault(require("path"));
const office_addin_usage_data_1 = require("office-addin-usage-data");
/* global process __dirname */
// File path defaults
exports.azCliInstallCommandPath = path_1.default.resolve(`${__dirname}/scripts/azCliInstallCmd.ps1`);
exports.azRestAddTenantOutlookReplyUrlsCommandPath = path_1.default.resolve(`${__dirname}/scripts/azRestAddTenantOutlookReplyUrls.txt`);
exports.azRestAddTenantReplyUrlsCommandPath = path_1.default.resolve(`${__dirname}/scripts/azRestAddTenantReplyUrls.txt`);
exports.azRestAddSecretCommandPath = path_1.default.resolve(`${__dirname}/scripts/azAddSecretCmd.txt`);
exports.azRestAppCreateCommandPath = path_1.default.resolve(`${__dirname}/scripts/azRestAppCreateCmd.txt`);
exports.azRestGetOrganizationDetailsCommandPath = path_1.default.resolve(`${__dirname}/scripts/azGetOrganizationDetails.txt`);
exports.azRestGetTenantAdminMembershipCommandPath = path_1.default.resolve(`${__dirname}/scripts/azRestGetTenantAdminMembership.txt`);
exports.azRestGetTenantRolesPath = path_1.default.resolve(`${__dirname}/scripts/azRestGetTenantRoles.txt`);
exports.azRestSetIdentifierUriCommmandPath = path_1.default.resolve(`${__dirname}/scripts/azRestSetIdentifierUri.txt`);
exports.azRestSetSigninAudienceCommandPath = path_1.default.resolve(`${__dirname}/scripts/azSetSignInAudienceCmd.txt`);
exports.envDataFilePath = path_1.default.resolve(`${process.cwd()}/.ENV`);
exports.fallbackAuthDialogTypescriptFilePath = path_1.default.resolve(`${process.cwd()}/src/helpers/fallbackAuthDialog.ts`);
exports.fallbackAuthDialogJavascriptFilePath = path_1.default.resolve(`${process.cwd()}/src/helpers/fallbackAuthDialog.js`);
exports.getInstalledAppsPath = path_1.default.resolve(`${__dirname}/scripts/getInstalledApps.ps1`);
exports.manifestFilePath = path_1.default.resolve(`${process.cwd()}/manifest.xml`);
exports.addSecretCommandPath = path_1.default.resolve(`${__dirname}/scripts/addAppSecret.ps1`);
exports.getSecretCommandPath = path_1.default.resolve(`${__dirname}/scripts/getAppSecret.ps1`);
exports.testEnvDataFilePath = path_1.default.resolve(`${process.cwd()}/test/test-env`);
exports.testFallbackAuthDialogFilePath = path_1.default.resolve(`${process.cwd()}/test/test-fallbackauthdialog`);
exports.testManifestFilePath = path_1.default.resolve(`${process.cwd()}/test/test-manifest.xml`);
// Usage data defaults
exports.usageDataObject = new office_addin_usage_data_1.OfficeAddinUsageData({
    projectName: "office-addin-sso",
    instrumentationKey: office_addin_usage_data_1.instrumentationKeyForOfficeAddinCLITools,
    raisePrompt: false,
});
