// office-addin-sso/lib/server.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SSOService = void 0;
// Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license. See full license in the root of the repo.
/*
    This file provides server configuration, startup and stop
*/
const https_1 = __importDefault(require("https"));
const devCerts = __importStar(require("office-addin-dev-certs"));
const office_addin_manifest_1 = require("office-addin-manifest");
const app_1 = require("./middle-tier/app");
const ssoDataSettings_1 = require("./ssoDataSettings");
const defaults_1 = require("./defaults");
/* global console process require */
require("dotenv").config();
class SSOService {
    constructor(manifestPath) {
        this.port = process.env.PORT || "3000";
        this.app = new app_1.App(this.port);
        this.app.initialize();
        this.manifestPath = manifestPath;
        this.ssoServiceStarted = false;
    }
    startSsoService(isTest = false) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                if (isTest) {
                    process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
                }
                yield this.getSecret(isTest);
                this.ssoServiceStarted = yield this.startServer(this.app.appInstance, this.port);
                defaults_1.usageDataObject.reportSuccess("startSsoService()");
                return Promise.resolve(this.ssoServiceStarted);
            }
            catch (err) {
                console.error(`Failed to start SSO server. ${err}`);
                defaults_1.usageDataObject.reportException("startSsoService()", err);
                return Promise.reject(false);
            }
        });
    }
    getSecret(isTest = false) {
        return __awaiter(this, void 0, void 0, function* () {
            const manifestInfo = yield office_addin_manifest_1.OfficeAddinManifest.readManifestFile(this.manifestPath);
            const appSecret = (0, ssoDataSettings_1.getSecretFromCredentialStore)(manifestInfo.displayName, isTest);
            if (appSecret === "") {
                const errorMessage = `Credential store does not have secret for '${manifestInfo.displayName}'.`;
                throw new Error(errorMessage);
            }
            process.env.secret = appSecret;
        });
    }
    getTestServerState() {
        return this.ssoServiceStarted;
    }
    startServer(app, port) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const options = yield devCerts.getHttpsServerOptions();
                this.server = https_1.default
                    .createServer(options, app)
                    .listen(port, () => console.log(`Server running on ${port}`));
                return Promise.resolve(true);
            }
            catch (err) {
                const errorMessage = `Unable to start test server on port ${port}: \n${err}`;
                return Promise.reject(errorMessage);
            }
        });
    }
    stopServer() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.ssoServiceStarted) {
                try {
                    this.server.close();
                    this.ssoServiceStarted = false;
                    return Promise.resolve(true);
                }
                catch (err) {
                    const errorMessage = `Unable to stop test server: \n${err}`;
                    return Promise.reject(new Error(errorMessage));
                }
            }
            else {
                // test server not started
                return Promise.resolve(false);
            }
        });
    }
}
exports.SSOService = SSOService;
